File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/OauthResource.java
package work.socialhub.kmastodon.api;

import mastodon4j.entity.AccessToken;
import mastodon4j.entity.share.Response;
import mastodon4j.internal._ListsResource;

/**
 * @author hecateball
 */
public interface OauthResource {

    /**
     * Attempt to login with the given credentials, and then retrieve the access token for the current user.
     *
     * @param clientId
     * @param clientSecret
     * @param emailAddress
     * @param password
     * @param scopes
     * @return an AccessToken
     */
    public Response<AccessToken> issueAccessToken(String clientId, String clientSecret, String emailAddress, String password, String scopes);

    /**
     * Attempt to login with authorization code, and then retrieve the access token for the current user.
     *
     * @param clientId
     * @param clientSecret
     * @param redirectUri
     * @param code
     * @return an AccessToken
     */
    public Response<AccessToken> issueAccessToken(String clientId, String clientSecret, String redirectUri, String code);

    /**
     * Attempt to renew access token with refresh token, and then retrieve new access token for the current user.
     *
     * @param clientId
     * @param clientSecret
     * @param refreshToken
     * @param refreshToken
     * @return an AccessToken
     */
    public Response<AccessToken> refreshAccessToken(String clientId, String clientSecret, String refreshToken);

    /**
     * Get Authorization URL
     *
     * @param clientId
     * @param redirectUri
     * @param scopes
     * @return an Authorization URL
     */
    public String getAuthorizationUrl(String clientId, String redirectUri, String scopes);
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/OAuthResource.kt
package work.socialhub.kmastodon.api

import work.socialhub.kmastodon.api.request.oauth.OAuthAuthorizationUrlRequest
import work.socialhub.kmastodon.api.request.oauth.OAuthIssueAccessTokenWithAuthorizationCodeRequest
import work.socialhub.kmastodon.api.request.oauth.OAuthIssueAccessTokenWithCredentialsRequest
import work.socialhub.kmastodon.api.request.oauth.OAuthRefreshAccessTokenRequest
import work.socialhub.kmastodon.api.response.Response
import work.socialhub.kmastodon.api.response.oauth.OAuthAuthorizationUrlResponse
import work.socialhub.kmastodon.api.response.oauth.OAuthIssueAccessTokenWithAuthorizationCodeResponse
import work.socialhub.kmastodon.api.response.oauth.OAuthIssueAccessTokenWithCredentialsResponse
import work.socialhub.kmastodon.api.response.oauth.OAuthRefreshAccessTokenResponse
import kotlin.js.JsExport

@JsExport
interface OAuthResource {

    /**
     * Attempt to log in with the given credentials,
     * and then retrieve the access token for the current user.
     */
    fun issueAccessTokenWithCredentials(
        request: OAuthIssueAccessTokenWithCredentialsRequest
    ): Response<OAuthIssueAccessTokenWithCredentialsResponse>

    /**
     * Attempt to log in with authorization code,
     * and then retrieve the access token for the current user.
     */
    fun issueAccessTokenWithAuthorizationCode(
        request: OAuthIssueAccessTokenWithAuthorizationCodeRequest
    ): Response<OAuthIssueAccessTokenWithAuthorizationCodeResponse>

    /**
     * Attempt to renew access token with refresh token,
     * and then retrieve new access token for the current user.
     */
    fun refreshAccessToken(
        request: OAuthRefreshAccessTokenRequest
    ): Response<OAuthRefreshAccessTokenResponse>

    /**
     * Get Authorization URL
     */
    fun authorizationUrl(
        request: OAuthAuthorizationUrlRequest
    ): Response<OAuthAuthorizationUrlResponse>
}
